gatsby-theme-auth0/src/hooks/useAuth.ts

```diff
@@ -1,41 +1,27 @@
 import * as React from "react";
-import auth, { SessionState } from "../auth/service";
-
-const useAuth = (stateCallback = (_state: SessionState) => {}) => {
-  const [isLoading, setIsLoading] = React.useState(true);
-  const [isLoggedIn, setIsLoggedIn] = React.useState(auth.isAuthenticated());
-  const [profile, setProfile] = React.useState(auth.getUserProfile());
+import auth, { AuthCallback } from "../auth/service";
+import { User, isLoggedInUser } from "../auth/user";
+import { SessionContext, Session } from "../components/SessionProvider";
 
+/**
+ * React hook for authentication.
+ *  - fetches the current logged-in state from the Session context provider
+ *  - todo: clarify if callback form is necessary
+ * @param stateCallback
+ * @returns { isLoading: boolean, user: User }
+ */
+const useAuth = (
+  stateCallback: AuthCallback = (_state: User) => {},
+): Session => {
+  const session = React.useContext(SessionContext);
   React.useEffect(() => {
-    // Override `sessionStateCallback` in auth service
-    auth.sessionStateCallback = state => {
-      stateCallback(state);
-      setIsLoggedIn(state.isLoggedIn);
-    };
-
-    (async () => {
-      await auth.checkSession();
-      try {
-        const user = auth.getUserProfile();
-        setProfile(user);
-      } catch (error) {
-        console.log(`Error: ${error}`);
-      }
-
-      setIsLoading(false);
-    })();
-
-    return () => {
-      // Clean up sessionStateCallback
-      auth.sessionStateCallback = () => {};
-    };
+    if (isLoggedInUser(session.user)) {
+      stateCallback(session.user);
+    } else {
+      // session.login(); // I don't think this is intended behavior; revisit
+    }
```

Comment:
This is one of the more complex pieces, so I'm sorry if I missed something about how it is intended to work (apparently I started documenting it above and got distracted).

Because the SessionProvider performs `auth.checkSession()` at boot time and makes it available via the context, I was able to remove that bit of complexity, but then I wasn't left with much to do- it essentially calls `useContext(SessionContext)`, calls its callback with that context, and then returns the session context as well. You can see it in action back in the [updates i made to the minimal demo (end of walkthrough)](https://github.com/epilande/gatsby-theme-auth0/pull/31/files#diff-0ab57148f4b017fb24e320548ce69d21R13-R31).

_One minor note: If multiple components on a page were using this hook, I was concerned they might conflict with each other since they are both potentially mutating auth's `sessionStateCallback`. This should no longer be an issue (and maybe it never was) but I wanted to point it out._